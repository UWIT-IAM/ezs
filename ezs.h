/* ========================================================================
 * Copyright (c) 2006 The University of Washington
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ========================================================================
 */

#ifndef EZS_H
#define EZS_H

/* Definitions from configure. */

#define EZS_NAME "ezs"
#define EZS_VERSION_STRING "ezs 1.4.2"
#define EZS_VERSION "1.4.2"

/* end of configure generated data */


/* EZS library definitions */


#if defined(WIN32)
#include <windows.h>
#else
#include <sys/errno.h>
#endif 

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include <string.h>

#include <openssl/bio.h>
#include <openssl/err.h>
#include <openssl/pem.h>
#include <openssl/evp.h>
#include <openssl/hmac.h>
#include <openssl/rand.h>
#include <openssl/ssl.h>
#include <openssl/x509.h>
#include <openssl/x509v3.h>

extern int ezs_debug;   /* set nonzero for debug to stderr */

/* structure for certs and keys */

typedef struct EZSCert_ {
  struct EZSCert_ *next;
  char *cn;
  X509 *x;
  EVP_PKEY *k;
} EZSCert;

typedef struct EZSCrl_ {
  struct EZSCrl_ *next;
  char *cn;
  X509_CRL *x;
} EZSCrl;

typedef struct EZSCtx_ {
  struct EZSCtx_ *next;
  SSL_CTX *ssl_ctx;
  int ref_count;
} EZSCtx;


/* The EZS structure includes openssl structures */

typedef struct EZS_ {
   EZSCtx *ezs_ctx;
   BIO *conn;
   SSL *ssl;
   SSL_SESSION *session;
   char *peer_hn;      /* hostname of peer - if known */
   char *peer_cn;      /* CN from peer's cert */
   char *cipher_list;  /* List of available ciphers */
   int nonblocking;    /* use non-blocking io */
   int timeout;        /* timeout seconds for semi-blocking io */
   int reinit;         /* set to reinitialize on each crypt */
   int requir_peer_cn; /* do we reqyur the peer's cert name? */
   int verify_peer_cn; /* do we verify the peer's cert name? */
   int allow_v2;       /* do we allow sslv2? */
   int allow_v3;       /* do we allow sslv3? */
   EVP_CIPHER_CTX *ct[2]; /* cryption context */
   EVP_CIPHER *crypt_cipher;
   EVP_MD *crypt_hash;
   int ezs_errno;      /* per-connection errno */
   int read_waiton_read;
   int read_waiton_write;
   int write_waiton_read;
   int write_waiton_write;
   int connecting;     /* connect or accept in progress */
} EZS;

#define EZS_DEFAULT_TIMEOUT 30 /* 30 sec */

/* cryption */
#define EZS_ENCRYPT 1  /* EVP library conventions */
#define EZS_DECRYPT 0
#define EZS_BLOCKSIZE(E) EVP_CIPHER_CTX_block_size(E->ct[0])

/* errors */

#define EZS_ERR_NONE      0
#define EZS_ERR_SSL       1
#define EZS_ERR_PEERCRT   2
#define EZS_ERR_PEERCN    3
#define EZS_ERR_NOCRT     4
#define EZS_ERR_BADCRT    5
#define EZS_ERR_BADKEY    6

#define EZS_ERR_NO_CONNECTION     7
#define EZS_ERR_READ_WOULDBLOCK   8
#define EZS_ERR_WRITE_WOULDBLOCK  9
#define EZS_ERR_READ_EOF         10
#define EZS_ERR_WRITE_EOF        11

#define EZS_ERR_INVALID          12
#define EZS_ERR_CANT_HAPPEN      13
#define EZS_ERR_MAX              14

#ifdef _EZS_LIBRARY_
char *ezs_err_txt[] = {
  "No error",
  "SSL error",
  "Peer cert is invalid",
  "Peer cert cn does not match peer hostname",
  "Cert not found",
  "Certificate is invalid",
  "Key is invalid",
  "No connection",
  "Read would block",
  "Write would block",
  "Read end of file",
  "Write end of file",
  "Invalid arguments",
  "Can't happen"
};
#endif



/* exported procedures */

#define ezs_set_cipher_list(E,t) (E->cipher_list=strdup(t))
#define ezs_set_peer_hn(E,t) (E->peer_hn=t)
#define ezs_set_require(E,t) (E->requir_peer_cn=t)
#define ezs_set_verify(E,t) (E->verify_peer_cn=t)
#define ezs_set_timeout(E,t) (E->timeout=t)
#define ezs_use_nonblocking_io(E) (E->nonblocking=1)
#define ezs_use_blocking_io(E) (E->nonblocking=0)
#define ezs_allow_v2(E) (E->allow_v2=1)
#define ezs_allow_v3(E) (E->allow_v3=1)
#define ezs_deny_v2(E) (E->allow_v2=0)
#define ezs_deny_v3(E) (E->allow_v3=0)
#define ezs_load_rand_file(f) RAND_load_file(f,-1);
#define ezs_save_rand_file(f) RAND_write_file(f);

int ezs_errno(EZS *E);
int ezs_geterror(EZS *E, char **txt);
int ezs_wouldblock(EZS *E);

int ezs_init(char *name);
EZS *ezs_new();
void ezs_free(EZS *L);
char *ezs_version();
int ezs_add_ca(char *cafile);
int ezs_add_crl(char *crlfile);
int ezs_add_cert(char *crtfile, char *keyfile, char **cn);
char *ezs_get_peer_cn(EZS *E);


void ezs_set_session_name(char *name);
char *ezs_get_session(EZS *L);
int ezs_set_session(EZS *L, char *session);

int ezs_connect(EZS *L, char *service, char *cn);
int ezs_connect_fd(EZS *L, int fd, char *cn);
int ezs_disconnect(EZS *L);
int ezs_listen(EZS *L, char *port);
EZS *ezs_accept(EZS *A, char *cn);
int ezs_accept_fd(EZS *A, int fd, char *cn);
EZSCert *ezs_find_cert(char *cn);

int ezs_write(EZS *L, char *txt, int len);
int ezs_read(EZS *L, char *txt, int len);

int ezs_crypt_init(EZS *L, char *keytext, int keylen, unsigned char **iv);
int ezs_crypt(EZS *L, int mode, unsigned char *out, int *outlen, unsigned char *in, int inlen);

int ezs_hmac(unsigned char **h, int *hlp, unsigned char *data, int dl, unsigned char *key, int kl);
int ezs_data_to_base64(char **b64, int *bl, void *data, int dl);
int ezs_base64_to_data( void **data, int *dl, char *b64, int bl);


#endif /* EZS_H */
