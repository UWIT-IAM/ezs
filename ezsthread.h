#ifndef EZSTHREAD_H
#define EZSTHREAD_H

#if defined(WIN32)
  #include <windows.h>
  #include <process.h>
#else
  #include <unistd.h>
  #include <pthread.h>
#endif

#include <openssl/ssl.h>

int ezs_THREAD_setup(void);
int ezs_THREAD_leanup(void);

#if defined(WIN32)

/* don't know about some of these */
    #define THREAD_T HANDLE
    #define THREAD_PROC void
    #define THREAD_CREATE(id, proc, arg) id=(HANDLE)_beginthread(proc,0,arg)
    #define THREAD_ID GetCurrentThreadId( )
    #define THREAD_DETACH 
    #define THREAD_CANCEL(t)  TerminateThread((HANDLE*)t, 0)
    #define THREAD_OKCANCEL 

    #define MUTEX_T HANDLE
    #define MUTEX_SETUP(x) (x) = CreateMutex(NULL, FALSE, NULL)
    #define MUTEX_CLEANUP(x) CloseHandle(x)
    #define MUTEX_LOCK(x) WaitForSingleObject((x), INFINITE)
    #define MUTEX_UNLOCK(x) ReleaseMutex(x)

#elif defined(_POSIX_THREADS)

    #define THREAD_T pthread_t
    #define THREAD_PROC void*
    #define THREAD_CREATE(id, proc, arg) pthread_create(&id, NULL, proc, (void*)arg)
    #define THREAD_ID pthread_self( )
    #define THREAD_DETACH pthread_detach(pthread_self());
    #define THREAD_CANCEL(t) pthread_cancel(t)
    #define THREAD_OKCANCEL pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL)

    #define MUTEX_T pthread_mutex_t
    #define MUTEX_SETUP(x) pthread_mutex_init(&(x), NULL)
    #define MUTEX_CLEANUP(x) pthread_mutex_destroy(&(x))
    #define MUTEX_LOCK(x) pthread_mutex_lock(&(x))
    #define MUTEX_UNLOCK(x) pthread_mutex_unlock(&(x))

#else
    #bomb No thread definitions!
#endif

#endif /* EZSTHREAD_H */
