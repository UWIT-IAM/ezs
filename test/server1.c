/* ========================================================================
 * Copyright (c) 2006 The University of Washington
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ========================================================================
 */

/*
   Server demo version 1
     use blocking io
     use ezs_accept

   Accept connection from client
     validate
     text from clients go to stdout
     text from stdin goes to clients
   */
 
#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#include <strings.h>
#include <sys/types.h>
#include <signal.h>

#include "ezs.h"
#include "ezsthread.h"

char *prog;
int debug = 1;  /* verbose */
int count_clients = 0;

int tot_clients = 0;
int tot_rt = 0;
int tot_wt = 0;
int tot_pipe = 0;

/** unused: ezs lib sets sigpipe to ignore 
void catch_pipe(int i)
{
   signal(SIGPIPE, catch_pipe);
   tot_pipe++;
}
 **/

static void show_client_count() 
{
   printf("\r%7d %4d %4d %3d", tot_clients, tot_rt, tot_wt, tot_pipe);
   fflush(stdout);
}

#ifdef WIN32
#include "openssl/applink.c"
#define sleep(s) Sleep(s*1000)
#endif

int usage()
{
   printf("usage: %s -p port [-q] [-d] -ca ca_pem [-crl crl_pem] -c server_cert [-k server_key] [-t timeout]\n", prog);
   exit (1);
}

void report_err(EZS *E, char *msg)
{
   int e;
   char *t;
   e = ezs_geterror(E, &t);
   if (debug||(e!=10)) fprintf(stderr, ".. %s, ezs error %d %s\n", msg, e, t);
   if (t) free(t);
}

void fatal_err(EZS *E, char *msg)
{
   report_err(E, msg);
   exit (1);
}


/* Because client reads and writes can block
   threads for each are created. Messages to clients are
   queued.  No such consideration is given to the console.
   */


/* Clients */

typedef struct String_ {
  struct String_ *next;
  char *str;
} String;

typedef struct Client_ {
  struct Client_ *next;
  int state;
  EZS *E;
  int no;
  String *msg;      /* messages to client */
  THREAD_T reader;
  THREAD_T writer;
} Client;

Client *clients = NULL;
int clientno = 0;
/* make it simple with just the one lock */
MUTEX_T client_lock;

Client *new_client(EZS *E)
{
   Client *C = (Client*) malloc(sizeof(Client));
   MUTEX_LOCK(client_lock);
   C->next = clients;
   C->state = 1;
   C->E = E;
   C->no = clientno++;
   C->msg = NULL;
   clients = C;
   MUTEX_UNLOCK(client_lock);
   return (C);
}

void free_client(Client *C)
{
   Client *c, *p;
   MUTEX_LOCK(client_lock);
   for (p=NULL,c=clients; c!=C; p=c,c=c->next);
   if (c) {
      if (p) p->next = c->next;
      else clients = c->next;
   }
   MUTEX_UNLOCK(client_lock);
   free(C);
}

/* add a message to a client's output - caller has client lock */
void add_client_msg(Client *C, char *str)
{
   String *S, *s;

   if (!C->state) return;

   s = (String*) malloc(sizeof(String));
   s->str = strdup(str);
   s->next = NULL;
   S = C->msg;
   while (S&&S->next) S = S->next;
   if (S) S->next = s;
   else C->msg = s;
}

char *get_client_msg(Client *C)
{
   char *v;
   String *S;

   MUTEX_LOCK(client_lock);
   S = C->msg;
   if (S) C->msg = S->next;
   MUTEX_UNLOCK(client_lock);

   if (!S) return (NULL);
   v = S->str;
   free(S);
   return (v);
}

/* client reader thread */

THREAD_PROC client_reader(void *arg)
{
    Client *C = (Client*) arg;
    EZS *E = C->E;
    char rbuf[4096];
    int nb;

    THREAD_DETACH;
    THREAD_OKCANCEL;
    C->state++;
    tot_rt++;
    if (debug>1) fprintf(stderr, "new client_thread\n");
    
    while (C->state && ((nb=ezs_read(E, rbuf, 4095))>0)) { 
        rbuf[nb] = '\0';
        if (debug) printf("[%d] recv %d bytes: %s\n", C->no, nb, rbuf);
        if (!strncmp(rbuf, "exit", 4)) exit (0);
    }

    if ( nb<=0) report_err(E, "read");
  
    /* THREAD_CANCEL(C->writer); */

    if (C->state==0) {
       ezs_disconnect(E);
       free_client(C);
    } else C->state = 0;
    
    tot_rt--;

}

/* client writer thread */

THREAD_PROC client_writer(void *arg)
{
    Client *C = (Client*) arg;
    EZS *E = C->E;
    char *msg;
    int rc;
    int no = C->no;

    THREAD_DETACH;
    THREAD_OKCANCEL;
    C->state++;
    tot_wt++;

    if (debug>1) fprintf(stderr, "new client_writer\n");

    while (C->state) {
       msg = get_client_msg(C);
       if (msg) {
          rc = ezs_write(E, msg, strlen(msg));
          if (debug) printf("[%d] sent %d bytes: %s\n", C->no, rc, msg);
          if (rc<=0) {
              report_err(E, "write");
              if (ezs_geterror(E, NULL)==EZS_ERR_NO_CONNECTION) break;
          }
       } else sleep(1);
    }

    if (C->state==0) {
       ezs_disconnect(E);
       free_client(C);
    } else C->state = 0;

    tot_wt--;
}

/* thread to receive input from the terminal */

THREAD_PROC user_thread(void *arg)
{
   char sdata[4096];
   Client *C;

   THREAD_DETACH;
   if (debug>1) fprintf(stderr, "new user_thread\n");

   for (;;) {
      if (!fgets(sdata, 4095, stdin)) {
          perror("stdin");
          exit (1);
      }
      if (debug>1) fprintf(stderr,"term: %s\n", sdata);
      
      if (!strncmp(sdata,"crl ",4)) {
         char *p, *e;
         for (p=sdata+4; *p==' '; p++);
         if (e=strchr(p, '\n')) *e = '\0';
         ezs_add_crl(p);
         
      } else if (!strcmp(sdata,"verbose")) {
         debug = 1;

      } else if (!strcmp(sdata,"debug")) {
         debug = 2;

      } else if (!strcmp(sdata,"quiet")) {
         debug = 0;

      } else if (!strcmp(sdata,"ncc")) {
         count_clients = 1;

      } else if (!strcmp(sdata,"cc")) {
         count_clients = 1;
         show_client_count();

      } else {  /* send message to clients */

         MUTEX_LOCK(client_lock);
         for (C=clients; C; C=C->next) {
            if (debug>1) fprintf(stderr,"add to %d: %s\n", C->no, sdata);
            add_client_msg(C, sdata);
         }
         MUTEX_UNLOCK(client_lock);
      }
      if (count_clients) show_client_count();
  } 
  exit (1);
}

main(argc,argv)
int  argc;
char **argv;
{
 
   pthread_t tid;
   char sdata [4096];
   int len;
   char *port = NULL;
   char *kfile = NULL;
   char *cfile = NULL;
   char *cafile = NULL;
   char *crlfile = NULL;
   char *cn = NULL;
   int timeout = 0;

   EZS *E;

   int term_input = 1;
    
   /* parse the args */

  /* signal(SIGPIPE, catch_pipe); */

  prog = argv[0];
  while (--argc) {
     argv++;
     if (!strcmp(argv[0], "-p")) {
        if (--argc<0) usage();
        port = (++argv)[0];
     } else if (!strcmp(argv[0], "-c")) {
        if (--argc<0) usage();
        cfile = (++argv)[0];
     } else if (!strcmp(argv[0], "-k")) {
        if (--argc<0) usage();
        kfile = (++argv)[0];
     } else if (!strcmp(argv[0], "-ca")) {
        if (--argc<0) usage();
        cafile = (++argv)[0];
     } else if (!strcmp(argv[0], "-crl")) {
        if (--argc<0) usage();
        crlfile = (++argv)[0];
     } else if (!strcmp(argv[0], "-q")) {
        debug = 0;
        ezs_debug = 0;
     } else if (!strcmp(argv[0], "-d")) {
        debug = 2;
        ezs_debug = 1;
     } else if (!strcmp(argv[0], "-nt")) {
        term_input = 0;
     } else if (!strcmp(argv[0], "-t")) {
        if (--argc<0) usage();
        timeout = atoi((++argv)[0]);
     } else if (!strcmp(argv[0], "-cc")) {
        count_clients = 1;
     } else usage();
  }

  if (!(port&&cfile&&cafile)) {
      fprintf(stderr,">> Need cert, ca, and port\n");
      usage();
  }
  if (!kfile) kfile = cfile;

  ezs_init("server1_demo");

  MUTEX_SETUP(client_lock);

  E = ezs_new();
  if (timeout>0) ezs_set_timeout(E, timeout);

  if (!ezs_add_ca(cafile)) fatal_err(NULL, "ca");
  if (crlfile && !ezs_add_crl(crlfile)) fatal_err(NULL, "crl");
  if (!ezs_add_cert(cfile, kfile, &cn)) fatal_err(NULL, "crt");
  if (debug) fprintf(stderr, "mycert cn=%s\n", cn);

  if (!ezs_listen(E, port)) fatal_err(NULL, "listen");

  printf("Accepting connections, port '%s'.\n", port);
  
  /* start a thread to receive user input */

  if (term_input) THREAD_CREATE(tid, user_thread, (void *)NULL);
 
/* Wait for connections */

  for (;;) {
  
     EZS *N;
     Client *C;

     N = ezs_accept(E, cn);

     if (!N) fatal_err(NULL, "accept");

     /* start readers and writer for this client */

     C = new_client(N);

     THREAD_CREATE(C->reader, client_reader, (void *)C);
     THREAD_CREATE(C->writer, client_writer, (void *)C);

     tot_clients++;
     if (count_clients) show_client_count();

  } 
  printf("Exiting\n");
  exit (1);
}



