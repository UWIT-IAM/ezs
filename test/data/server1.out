mycert cn=localhost
.. read, ezs error 10 Read end of file
.. read, ezs error 10 Read end of file
.. read, ezs error 10 Read end of file
Accepting connections, port '98765'.
[2] recv 26 bytes: Hello, world. First time.

[2] recv 15 bytes: Goodbye world.

[3] recv 27 bytes: Hello, world. Second time.

[3] recv 5 bytes: exit

